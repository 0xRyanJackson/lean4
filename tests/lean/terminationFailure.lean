def f (x : Nat) : Nat :=
  if h : x > 0 then
    g x + 2
  else
    1
where
  g : Nat â†’ Nat
  | 0 => 2
  | x => f x * 2

#check! f
#check! f.g

/- Uncomment after stage0
#eval! f 0
#eval! f.g 0
-/

inductive Foo where
  | a | b | c
  deriving Repr
def h (x : Nat) : Foo :=
  match x with
  | 0 => Foo.a
  | 1 => Foo.b
  | x => h x

#check h

/- Uncomment after stage0
#eval h 0
-/
