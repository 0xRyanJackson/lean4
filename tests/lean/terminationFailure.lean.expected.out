terminationFailure.lean:7:2-7:3: error: fail to show termination for
  f.g
  f
with errors
failed to infer structural recursion:
Cannot use parameters #1 of f.g and x of f:
  failed to eliminate recursive application
    _root_.f x


Could not find a decreasing measure.
The arguments relate at each recursive call as follows:
(<, ≤, =: relation proved, ? all proofs failed, _: no proof attempted)
Call from f.g to f at 9:9-12:
  x1
x  =
Call from f to f.g at 3:4-7:
   x
x1 =

Please use `termination_by` to specify a decreasing measure.
terminationFailure.lean:11:8-11:9: error: application type mismatch
  !f
argument
  f
has type
  Nat → Nat : Type
but is expected to have type
  Bool : Type
!sorryAx Bool true : Bool
terminationFailure.lean:12:8-12:11: error: application type mismatch
  !f.g
argument
  f.g
has type
  Nat → Nat : Type
but is expected to have type
  Bool : Type
!sorryAx Bool true : Bool
terminationFailure.lean:22:4-22:5: error: fail to show termination for
  h
with errors
failed to infer structural recursion:
Cannot use parameter x:
  failed to eliminate recursive application
    h x


failed to prove termination, possible solutions:
  - Use `have`-expressions to prove the remaining goals
  - Use `termination_by` to specify a different well-founded relation
  - Use `decreasing_by` to specify your own tactic for discharging this kind of goal
x : Nat
⊢ False
h (x : Nat) : Foo
